src/internal/pfsdb/common.go

```diff
@@ -0,0 +1,199 @@
+)
+
+type (
+	SortOrder      int
+	FilterOperator string
+)
+
+const (
+	SortNone SortOrder = iota
+	SortAscend
+	SortDescend
+)
+
+const (
+	// Note that we use `?` as the placeholder for bindvar, but postgres uses $1, $2, etc.
+	// We solve this by rebinding the query before executing it using sqlx.Rebind.
+	Equal              FilterOperator = "%s = ?"
+	GreaterThan        FilterOperator = "%s > ?"
+	GreaterThanOrEqual FilterOperator = "%s >= ?"
+	LessThan           FilterOperator = "%s < ?"
+	LessThanOrEqual    FilterOperator = "%s <= ?"
+	NotEqual           FilterOperator = "%s <> ?"
+	ValueIn            FilterOperator = "%s in (?)"
+)
+
+type (
+	ModelType interface{ Repo | Commit | Branch }
+	FieldType interface{ CommitField | BranchField }
+	PairType  interface{ CommitPair | BranchPair }
+)
+
+type Filter[T FieldType] struct {
+	Field  T
+	Value  any
+	values []any
+	Op     FilterOperator
+}
+
+func (f *Filter[T]) QueryString() (formatted string, err error) {
+	formatted = fmt.Sprintf(string(f.Op), f.Field)
+	if f.Op == ValueIn {
+		formatted, f.values, err = sqlx.In(formatted, f.Value) // Thank god for this!
+		if err != nil {
+			return "", errors.Wrapf(err, "expanding %s", formatted)
+		}
+	}
+	return
+}
+
+type AndFilters[T FieldType] []*Filter[T]
+
+func (f AndFilters[T]) QueryString() (string, error) {
+	var parts []string
+	for _, filter := range f {
+		part, err := filter.QueryString()
+		if err != nil {
+			return "", err
+		}
+		parts = append(parts, part)
+	}
+	return strings.Join(parts, " AND "), nil
+}
+
+type OrderBy[T FieldType] struct {
+	Fields    []T
+	SortOrder SortOrder
+}
+
+func (o *OrderBy[T]) QueryString() string {
+	var parts []string
+	for _, field := range o.Fields {
+		parts = append(parts, string(field))
+	}
+	var order string
+	switch o.SortOrder {
+	case SortAscend:
+		order = "ASC"
+	case SortDescend:
+		order = "DESC"
+	}
+	return strings.Join(parts, ", ") + " " + order
+}
+
+type QueryBuilder[T FieldType] struct {
+	baseQuery   string
+	queryParams []any
+
+	AndFilters AndFilters[T]
+	GroupBy    string
+	OrderBy    *OrderBy[T]
+	// TODO, should QueryBuilder deal with Limit and Offset?
+}
+
+func (c *QueryBuilder[T]) Build() (string, error) {
+	query := c.baseQuery
+	condition, err := c.AndFilters.QueryString()
+	if err != nil {
+		return "", err
+	}
+	if condition != "" {
+		for _, filter := range c.AndFilters {
+			if filter.Op == ValueIn {
+				// filter.values is set by Filter.String()
+				c.queryParams = append(c.queryParams, filter.values...)
+			} else {
+				c.queryParams = append(c.queryParams, filter.Value)
+			}
+		}
+		query += "\nWHERE " + condition
+	}
+	// TODO implement GROUP BY
+	if c.GroupBy != "" {
+		query += "\nGROUP BY " + c.GroupBy
+	}
+	if c.OrderBy != nil {
+		query += "\nORDER BY " + c.OrderBy.QueryString()
+	}
+	query = sqlx.Rebind(sqlx.DOLLAR, query)
+	return query, nil
+}
+
+type transformFn[T ModelType, U PairType] func(context.Context, *pachsql.Tx, *T) (*U, error)
```

Comment:
I think the API should aim to produce U with indirecting through an intermediate function.  Otherwise, you are doing one round trip per row, which is not ideal.